openfl-tutorials/experimental/Workflow_FedProx_Keras_MNIST.ipynb

```diff
@@ -0,0 +1,1304 @@
+     ]
+    }
+   ],
+   "source": [
+    "from openfl.experimental.interface import Collaborator, Aggregator\n",
+    "from openfl.experimental.runtime import LocalRuntime\n",
+    "\n",
+    "def callable_to_initialize_collaborator_private_attributes(n_collaborators, index, train_dataset, test_dataset, batch_size):\n",
+    "    train_splitter = LogNormalNumPyDataSplitter(num_clients=n_collaborators,num_classes=10, mu=0, sigma=1)\n",
+    "    test_splitter = LogNormalNumPyDataSplitter(num_clients=n_collaborators,num_classes=10, mu=0, sigma=1)\n",
+    "\n",
+    "\n",
+    "    X_train, y_train = train_dataset\n",
+    "    X_test, y_test = test_dataset\n",
+    "    train_idx = train_splitter.split(X_train, y_train)\n",
+    "    valid_idx = test_splitter.split(X_test, y_test)\n",
+    "\n",
+    "    train_dataset = X_train[train_idx[index]], y_train[train_idx[index]]\n",
+    "    test_dataset = X_test[valid_idx[index]], y_test[valid_idx[index]]\n",
+    "\n",
+    "    return {\n",
+    "        \"train_loader\": train_dataset, \"test_loader\": test_dataset,\n",
+    "        \"batch_size\": batch_size\n",
+    "    }\n",
+    "    \n",
+    "# # Setup participants\n",
+    "aggregator = Aggregator()\n",
+    "# aggregator.private_attributes = {}\n",
+    "collaborators = []\n",
+    "collaborator_names = ['Portland', 'Seattle', 'Chandler','Bangalore']\n",
+    "for idx, collaborator_name in enumerate(collaborator_names):\n",
+    "    collaborators.append(\n",
+    "        Collaborator(\n",
+    "            name=collaborator_name, num_cpus=0, num_gpus=0.3,\n",
+    "            private_attributes_callable=callable_to_initialize_collaborator_private_attributes,\n",
+    "            n_collaborators=len(collaborator_names), index=idx, train_dataset=(train_images, train_labels),\n",
+    "            test_dataset=(test_images, test_labels), batch_size=32\n",
+    "        )\n",
+    "    )\n",
+    "local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='single_process')\n",
+    "print(f'Local runtime collaborators = {local_runtime.collaborators}')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "436dff92",
+   "metadata": {},
+   "source": [
+    "We use the fedprox optimizer to train the model. The optimizer is defined in the following cell. (see https://arxiv.org/pdf/1812.06127.pdf).\n",
+    "The implementation is the same as in `openfl/utilities/optimizers/keras/fedprox.py`. Here, to support newer version of tensor flow we inherent from the legacy optimizer and add the build method."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 73,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tensorflow as tf\n",
+    "import tensorflow.keras as keras\n",
+    "from tensorflow.python.ops import standard_ops\n",
+    "\n",
+    "\n",
+    "@keras.utils.register_keras_serializable()\n",
+    "class FedProxOptimizer(keras.optimizers.legacy.Optimizer):\n",
```

Comment:
An alternative option here would be to implement FedProx via a custom loss function instead of at the optimizer level. 

However, as far as I know, keras FedProx is not used in any "live" tutorials or workflows, so it would be safe to upgrade the package directly too. Only caveat is that we will eventually need to update it for Keras3, sooner or later

Reply:
see my comment above.

Reply:
@yontyon, I agree with @kta-intel, so I'd suggest at least assessing how big would be the change of upgrading the [Keras FedProxOptimizer](https://github.com/securefederatedai/openfl/blob/develop/openfl/utilities/optimizers/keras/fedprox.py) at the framework level. If non-trivial, it can be done in a follow-up PR too.